src/runtime/relax_vm/paged_kv_cache.cc

```diff
@@ -0,0 +1,633 @@
+    last_page_offset_device_ = NDArray::Empty({reserved_num_seqs}, dtype_aux_, device);
+    cur_append_length_indptr_device_ = NDArray::Empty({reserved_num_seqs + 1}, dtype_aux_, device);
+    cur_pos2seqid_device_ = NDArray::Empty({reserved_num_pages * page_size}, dtype_aux_, device);
+  }
+
+  /*!
+   * \brief Given a sequence id and a required extra length, allocate new
+   * new pages for the sequence until the total capacity can cover the
+   * current sequence length plus the required extra length.
+   * \note By reserving extra length for the sequence, the subsequent appended
+   * K/V data for the sequence must have the same length as the input length here.
+   * \param seq_id The id of the sequence to process.
+   * \param extra_length The extra length to reserve for the sequence.
+   */
+  void ReserveExtraLengthForAppend(int64_t seq_id, int64_t extra_length) {
+    CHECK_GE(seq_id, 0) << "Input sequence id should be positive";
+    CHECK_LE(seq_id, num_total_seqs_)
+        << "Invalid input sequence id " << seq_id
+        << ", which is larger than the current total number of sequences.";
+    CHECK_GT(extra_length, 0) << "The input length should be positive.";
+
+    if (seq_id == num_total_seqs_) {
+      // Initialize if it is a new sequence.
+      page_table_.push_back({});
+      seq_lengths_.push_back(0);
+      cur_append_lengths_.push_back(0);
+      ++num_total_seqs_;
+    }
+
+    // The reservation is based on the current sequence length.
+    // If "current sequence + input extra length" does not exceed the
+    // current capacity (number of pages * page size), no action is taken.
+    int64_t cur_npage = page_table_[seq_id].size();
+    int64_t tgt_npage = (seq_lengths_[seq_id] + extra_length + page_size_ - 1) / page_size_;
+    for (int64_t page_idx = cur_npage; page_idx < tgt_npage; ++page_idx) {
+      AllocatePageForSequence(seq_id);
+    }
+    seq_lengths_[seq_id] += extra_length;
+    cur_append_lengths_[seq_id] = extra_length;
+    dirty_aux_data_device_ = true;
+  }
+
+  /*!
+   * \brief The entrance of attention. It takes an attention compute
+   * function together with the query data, invokes the attention
+   * function to complete the computation.
+   * \param f_attention The input attention compute function.
+   * \param q_data The query data. We support the following two layout settings:
+   * - in batch decode settings, q_data has layout (num_total_seqs, 1, num_heads, head_dim)
+   *   where num_total_seqs should exactly equal to the number of sequences in the cache.
+   * - in other settings (single-sequence prefill, batch prefill, or speculation
+   *   verification), q_data has the **flattened layout** to handle raggedness:
+   *   (1, total query length, num_heads, head_dim).
+   * \param layer_id The model layer index of the current attention.
+   * \param output The attention output array.
+   * \param apply_rotary A boolean flag indicating if to apply RoPE.
```

Comment:
Apply RoPE to the kv? or qkv, would be good to state the purpose

Reply:
It might be useful to add a TODO and indicate that we could move the rope related parameters to the f_attention closure in the future.